### PR TITLE
Refactor BDAddr

### PR DESCRIPTION
**Pull request by MattesWhite**
This PR follows the discussion of #68.

At last I found some time to work on this. For now I refactored `BDAddr` as a show case what I intend to do. If the implementation is approved I'll start working on `UUID`. Of course I'll include the changed suggested in #80 as well.

Initially, I wanted to write the parsers with `nom` but I recognized that there is a bunch of `nom v4` code in the crate so I settled for a 'hand-crafted' version. Another solution would be to use `regex` but I didn't wanted to pull in another dependency without asking.

### CODE REVIEW

**Comment by qdot** [comment]
I definitely like the direction this is going, though do note that your current version is failing on macOS and Windows.

I think you're fine to move on to UUID, though note that I did bring in #80 (as I'm about to do a quick point release to fix Windows UWP reads), but you can integrate that in as you need.

The `nom` dependency thing is killing me too, I'm really hoping to get back to this one day and possibly just move us to actual dbus usage. :(

I don't think bringing in `regex` would be a huge issue? But being an ex browser dev I'm also kind of oblivious to compile times and all that, and the library I use this for is... pretty violently monolithic so I have like 30+ dependencies there. :)

**Comment by MattesWhite** [comment]
The `BDAddr` part is now finalized. All pipelines should only fail because of legacy code by now.

A question of style: Do you prefer having bigger modules (`src/api/mod.rs` has now more than 500 lines without tests) or smaller once, e.g. put `BDAddr` and `UUID` into submodules and re-export them? Personally, I'd prefer the latter but I'll align with your choice of style.

Once, I did a refactoring from `nom v4` to `nom v5` so I'm willing to take care of this. However, the issue of having no RSSI values available (#76) is way more pressing for me. In addition, there is already worked on [`nom v6`](Geal/nom#1132) which probably will improve support for custom errors so waiting a little bit and solving #48 first is maybe not that bad.

For now I'll focus on refactoring UUIDs.

**Comment by qdot** [comment]
Looks like everything built on the latest commit, so that's positive.

And smaller modules, please! The big modules came in with the libraries I imported, I'm much more a fan of submodules and re-exporting.

For the nom stuff, it's really hard to say how much that matters other than the conflicting dependencies, especially if we start looking toward dbus again for bluez. The RSSI is now important for me too (My library that depends on this now advertises RSSI relay capabilities, and yet, I have no RSSI here... >.> ), so hopefully that can be looked at soon.

**Comment by MattesWhite** [comment]
So first work on UUIDs is done. I decided that its most practical to use [`uuid::Uuid`](https://docs.rs/uuid/0.8.1/uuid/struct.Uuid.html) wherever possible but I kept the short-value variant which has nice benefits like use parse and format functions of `uuid` crate.

I can't guarantee that the Windows version still works or that I didn't introduce bugs. I tried to adapt what was possible but when I the need to reverse bytes here and there makes my head hurt. In addition, I have only a Linux machine to test, so I don't even know if it compiles. Is there a way to do (at least) a `cargo check` for another OS? All in all, it would be really nice if someone could verify the code on a Windows machine.

**Comment by qdot** [comment]
After the issues I had with the macOS read/write patches, I'll be testing this throughly across all platforms before landing, because it directly affects users of my libraries that depend on this. 👍 

So far, according to CI it looks like the only thing catching you on windows is a &u16 to u32 miscast, so not *too* horrible?

**Comment by MattesWhite** [comment]
Yes the fixing errors caught by the pipelines is easy but I don't wanted to fix each error separately, push a single commit, wait for CI, iterate and so on. This also costs pipeline time, I think? What really bothers me is that I often read that Windows GUIDs are little endian and the `uuid::Uuid` is big endian so I can't guarantee that the conversion functions do the right thing. And writing tests without IDE support is not very pleasant.

**Comment by qdot** [comment]
Well, looks like you pass CI now at least. And I all do all my development on windows (and the vast number of users of this library are also on windows, and probably don't know they're using this library), so like I said, I'll at least give it a runthru before landing. :)

**Comment by johnmave126** [comment]
Actually, how about using [eui48](https://crates.io/crates/eui48) for mac address? That crate seems pretty slim and seems to be treated as de-facto mac address crate in crate like tokio-postgres.

**Comment by qwandor** [comment]
@MattesWhite Hi! Any chance you could rebase the BDAddr changes against the latest dev branch? The diff is a bit hard to follow at the moment.

I've done something a bit similar for UUIDs in #123, but using the type from the `uuid` crate directly in the API and internally, which makes equality work properly, and then using an extension trait and a helper function to convert to and from the short form where needed.

**Comment by NZSmartie** [comment]
I've rebased it on to the current `dev` branch and omitted all UUID related changes. It did drop the useful UUID constants you had defined, but I'm sure we could add that in another PR, that way this PR only focuses on `BDAddr`

File: src/corebluetooth/adapter.rs [hunk 1]
```diff
@@ -17,9 +17,8 @@ pub struct Adapter {
 }
 
 pub(crate) fn uuid_to_bdaddr(uuid: &String) -> BDAddr {
-    BDAddr {
-        address: uuid.as_bytes()[0..6].try_into().unwrap(),
-    }
+    let b: [u8; 6] = uuid.as_bytes()[0..6].try_into().unwrap();
+    BDAddr::try_from(b).unwrap()
```
Comment (NZSmartie) [inline src/corebluetooth/adapter.rs:21]: One of refactors that didn't quite survive the rebase was this. I may be worth reviewing as I ended up applying a dirty fix so that macOS/iOS would still build

File: src/winrtble/utils.rs [hunk 2]
```diff
@@ -78,37 +62,8 @@ pub fn to_char_props(_: &GattCharacteristicProperties) -> CharPropFlags {
...
-    #[test]
-    fn check_address() {
-        let bluetooth_address = 252566450624623;
-        let addr = to_addr(bluetooth_address);
-        let result = to_address(addr);
-        assert_eq!(bluetooth_address, result);
-    }
-
-    #[test]
-    fn check_uuid_to_guid_conversion() {
```
Comment (qwandor) [inline src/winrtble/utils.rs:94]: These UUID tests are still useful, please don't delete them.

Comment (NZSmartie) [reply src/winrtble/utils.rs:94]: oh right, missed them when the UUID code got moved into its own module

**Comment by MattesWhite** [comment]
Hi all,

sorry for the late response. Lately I don't have much time to monitor the progress of `btleplug`. However, I'm very exited to see the progress on `async` :+1: .

Feel free to edit this PR as much as you like. I'm happy that at least some pieces of my code end up in the crate. If there are any questions I'll be happy to help.

File: src/api/mod.rs [hunk 3]
```diff
@@ -207,8 +142,8 @@ pub struct Characteristic {
     pub properties: CharPropFlags,
 }
 
-impl Display for Characteristic {
-    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+impl fmt::Display for Characteristic {
```
Comment (qwandor) [inline src/api/mod.rs:145]: What's wrong with importing `Display`. `Debug`, `Formatter` and so on as we were before?

File: src/api/mod.rs [hunk 4]
```diff
@@ -385,28 +320,7 @@ pub trait Central: Send + Sync + Clone {
...
-            assert_eq!(result, expected);
-
-            if let Ok(uuid) = result {
-                assert_eq!(input, uuid.to_string());
-            }
-        }
-    }
+    // use super::*;
+
+    // No tests, yet.
```
Comment (qwandor) [inline src/api/mod.rs:325]: Just remove the tests module, rather than keeping it here empty.

File: src/lib.rs [hunk 5]
```diff
@@ -108,4 +108,4 @@ pub enum Error {
 }
 
 // BtlePlug Result type
-pub type Result<T> = result::Result<T, Error>;
+pub type Result<T, E = Error> = result::Result<T, E>;
```
Comment (qwandor) [inline src/lib.rs:111]: This seems like a weird change to make. If you want a result with a different error type somewhere, can't you just use `std::result::Result` there?

File: src/api/bdaddr.rs [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+use crate::Error;
+
+/// Stores the 6 byte address used to identify Bluetooth devices.
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_cr")
+)]
+#[derive(Copy, Clone, Hash, Eq, PartialEq, Default)]
+#[repr(C)]
```
Comment (qwandor) [inline src/api/bdaddr.rs:21]: Is this type being used for FFI somewhere?

File: src/api/bdaddr.rs [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+pub enum ParseBDAddrError {
+    #[error("Bluetooth address has to be 6 bytes long")]
+    IncorrectByteCount,
+    #[error("All digits in a Bluetooth address must be hex-digits [0-9a-fA-F]")]
+    InvalidDigit,
+}
+
+impl fmt::Display for BDAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        <Self as fmt::LowerHex>::fmt(self, f)
```
Comment (qwandor) [inline src/api/bdaddr.rs:36]: Uppercase is more common, can we keep that as the default?

File: src/api/bdaddr.rs [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+            Self::from_str_no_delim(s)
+        }
+    }
+}
+
+impl BDAddr {
+    pub fn into_inner(self) -> [u8; 6] {
+        self.address
+    }
+    pub fn bytes(&self) -> &[u8; 6] {
```
Comment (qwandor) [inline src/api/bdaddr.rs:148]: Nit: add a blank line between each method.

Comment (qwandor) [inline src/api/bdaddr.rs:148]: This method seems to be the same as `as_ref` above. No need for both.

File: src/api/bdaddr.rs [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+
+        let mut address = [0; 6];
+        for i in (0..12).step_by(2) {
+            let part = &s[i..i + 2];
+            address[i / 2] = u8::from_str_radix(part, 16).expect("Checked upfront");
+        }
+        Ok(Self { address })
+    }
+    /// Writes the address without delimiters.
+    pub fn write_flat(&self, f: &mut impl fmt::Write) -> fmt::Result {
```
Comment (qwandor) [inline src/api/bdaddr.rs:193]: Is this actually used anywhere? I'd be inclined to leave it and `to_string_flat` out unless you have a usecase in mind.

File: src/api/bdaddr.rs [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        if s.len() != 12 {
+            return Err(ParseBDAddrError::IncorrectByteCount);
+        }
+        if s.bytes().any(|b| !b.is_ascii_hexdigit()) {
+            return Err(ParseBDAddrError::InvalidDigit);
+        }
+
+        let mut address = [0; 6];
+        for i in (0..12).step_by(2) {
+            let part = &s[i..i + 2];
```
Comment (qwandor) [inline src/api/bdaddr.rs:187]: You can probably do this more nicely with `s.chunks(2)`.

File: src/api/bdaddr.rs [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+            address[i / 2] = u8::from_str_radix(part, 16).expect("Checked upfront");
```
Comment (qwandor) [inline src/api/bdaddr.rs:188]: Rather than checking twice, remove the `is_ascii_hexdigit` check above, and return the error here instead.

File: src/api/bdaddr.rs [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+            cpy.copy_from_slice(&slice[..6]);
+            Ok(cpy.into())
+        }
+    }
+}
+
+impl From<u64> for BDAddr {
+    fn from(int: u64) -> Self {
+        let mut cpy = [0; 6];
+        let slice = int.to_be_bytes(); // Reverse order to have MSB on index 0
```
Comment (qwandor) [inline src/api/bdaddr.rs:108]: Check that the top two bytes are 0.

**Comment by qwandor** [comment]
I think that this is generally an improvement, once the a bunch of small things are fixed. However, we could consider instead using one of the existing crates which provides a standard MAC address type, such as [eui48](https://crates.io/crates/eui48) or [macaddr](https://crates.io/crates/macaddr). I haven't had a chance yet to see which of the two is a better choice.

File: src/api/mod.rs (hunk 3 above)
Comment (MattesWhite) [reply src/api/mod.rs:145]: That's just personal preference, rather use the module name than importing everything. Especially, as we have to use `fmt::Result` anyway.

However, if this is contrary to the common style there is no problem in importing stuff from `std::fmt` directly.

File: src/lib.rs (hunk 5 above)
Comment (MattesWhite) [reply src/lib.rs:111]: Again just personal preference, I have started a [thread in URLO](https://users.rust-lang.org/t/dont-do-type-result-t-std-result-t-myerror/49421) about that topic. tl;dr Why should I be required to use `std::result::Result` just to state that other type parameters are used.

... back when I wrote the PR I was not that aware of 'keep a PR small and dedicated to a problem and keep personal preference out' or however it is expressed best (I'm not an english native speaker).

File: src/api/bdaddr.rs (hunk 6 above)
Comment (MattesWhite) [reply src/api/bdaddr.rs:21]: That is legacy from rumble (?). I wouldn't mind removing it.

File: src/api/bdaddr.rs (hunk 8 above)
Comment (MattesWhite) [reply src/api/bdaddr.rs:148]: Or a comment would be even better :grin:

Comment (MattesWhite) [reply src/api/bdaddr.rs:148]: Yeah, if I remember correctly, this was a legacy function but if we are rewriting the whole thing anyway, we can remove it as well.

File: src/api/bdaddr.rs (hunk 9 above)
Comment (MattesWhite) [reply src/api/bdaddr.rs:193]: Actually this was my initial motivation to write this PR (see https://github.com/deviceplug/btleplug/issues/68#issuecomment-666211155). I'm working on a 'Bluetooth gateway' where the BDAddr of devices is used as the identifier in a REST API. As the default `:` in MAC addresses is also a valid separator in URIs I use this _flat_ format. `eui48` provides a (similar format](https://docs.rs/eui48/1.1.0/eui48/struct.MacAddress.html#method.to_hexadecimal).

Don't know if this is enough to justify its existence in `btleplug`.

File: src/api/bdaddr.rs (hunk 12 above)
Comment (MattesWhite) [reply src/api/bdaddr.rs:108]: Is this really a problem/error? Would you rather use `TryFrom`?

File: src/api/bdaddr.rs (hunk 9 above)
Comment (MattesWhite) [reply src/api/bdaddr.rs]: At least I renamed it to `_no_delim` instead of `_flat`

File: src/lib.rs (hunk 5 above)
Comment (qwandor) [reply src/lib.rs:114]: It looks like it isn't used anymore anyway.

File: src/winrtble/utils.rs (hunk 2 above)
Comment (qwandor) [reply src/winrtble/utils.rs:94]: I've added them back.

File: src/api/bdaddr.rs (hunk 12 above)
Comment (qwandor) [reply src/api/bdaddr.rs:104]: I think that's better. Done this myself.
